### ISSUE TITLE
MulitProc Workflows locking up

### ISSUE THREAD

**Issue by blakedewey** [nipy/nipype#1692]
I think this may be a result of #1689. When I run a large MultiProc workflow, it runs for one or two nodes, then hangs. There is nothing reported and the nodes have finished successfully. I am also unable to use SIGTERM to end the process and have to manually close the terminal window. Moving back before this change seems to fix the issue.

**Comment by satra** [comment]
@carolFrohlich - could you please take a look at this?

**Comment by chrisfilo** [comment]
CCing @ccraddock who promised to help. We have experienced those as well. Deadlocks are stochastic in nature so there are no steps to reproduce. Here are nipype tests which timed out due to deadlocks in multiproc:

https://circleci.com/gh/nipy/nipype/1690
https://circleci.com/gh/nipy/nipype/1686
https://circleci.com/gh/nipy/nipype/1686
https://circleci.com/gh/nipy/nipype/1685
https://circleci.com/gh/nipy/nipype/1682
https://circleci.com/gh/nipy/nipype/1676
https://circleci.com/gh/nipy/nipype/1674

the problem seems to affect Python 2.7 more often than python 3.5

**Comment by chrisfilo** [comment]
It seems to be affecting travis tests as well:

https://api.travis-ci.org/jobs/180558248/log.txt?deansi=true
```
nipype.pipeline.plugins.tests.test_linear.test_run_in_series([1, 1], [1, 1]) ... ok


No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received

The build has been terminated
```

On a [successful run](https://api.travis-ci.org/jobs/179964049/log.txt?deansi=true) the next test after `test_run_in_series` is `nipype.pipeline.plugins.tests.test_multiproc.test_run_multiproc`

**Comment by djarecka** [comment]
As far as I remember my problems with travis were always in the second run only (python2.7; INSTALL_DEB_DEPENDECIES=false), e.g. https://travis-ci.org/djarecka/nipype/jobs/180111821
Checking now my CircleCI builds I also can't find problems with python3.

**Comment by jvarada** [comment]
Is there a reason for using the multiprocessing library and handling locks for the main thread manually rather than using concurrent.futures (or futures in py2.7), sending threads out as Futures, and waiting on the Futures instead?

**Comment by chrisfilo** [comment]
@ccraddock is [asking for example pipelines](https://github.com/nipy/nipype/pull/1739#issuecomment-265563799) that experienced this problem. Please add to the list.

The fmri_spm_nested pipeline included in nipype has experienced this issue at least a couple of times: https://circleci.com/gh/nipy/nipype/1690
https://circleci.com/gh/nipy/nipype/1674

We have seen it while running [fmriprep](https://github.com/poldracklab/fmriprep) as well (before switching to Python 3).

Unfortunately this issue seems to be stochastic in nature and is not easily reproducible.

I hope this helps

**Comment by satra** [comment]
@jvarada - mostly historical and our older support for 2.6. this would be a good thing to reconsider as we rewrite the engine.

**Comment by djarecka** [comment]
Not sure if you're testing on OSX, but I've just run locally and have problem with test_multiproc.py, python2.7.

@chrisfilo did you have any specific reason why you made skipif using 'TRAVIS_PYTHON_VERSION', i.e. `os.environ.get('TRAVIS_PYTHON_VERSION') == '2.7'`? Is it ok to change this using `sys.version_info`? Just trying to eliminate this problem when running `make tests`

**Comment by chrisfilo** [comment]
@djarecka There was no specific reason. Using `sys.version_info` seems more robust.

**Comment by djarecka** [comment]
@chrisfilo ok, will add an extra commit to pytest PR

**Comment by ccraddock** [comment]
@djarecka We found some interactions with nosetest that we believe we fixed here: https://github.com/nipy/nipype/pull/1739. Would you be able to see if these changes fix your problems?

**Comment by djarecka** [comment]
@ccraddock well, it will be hard to say since the problem doesn't occur every time. I'll pull your PR later and try to run a few times.

**Comment by ccraddock** [comment]
@djarecka We have figured out a way to make this problem easier to reproduce with nosetests. The problem is most likely to occur the first time you run the nosetest. If it hangs and you kill it, then it is very likely that it will work the second time. But if you wait approx 10min (maybe more) it will happen again. In our experience, you _have_ to delete the workflow directory (called pipe) in order to reproduce the problem.

**Comment by chrisfilo** [comment]
@ccraddock The PR also does not touch any of the MultiProc plugin code (only the tests) so I don't see how it could fix the core of the problem (causing deadlocks when running workflows - see example above). What am I getting wrong?

**Comment by djarecka** [comment]
@ccraddock I'm not completely sure, but I think the problem I had locally on my OSX today was not during the first run (within 10min).  FYI, I run pytest, not nosetest

**Comment by ccraddock** [comment]
Thanks for the info

**Comment by djarecka** [comment]
@chrisfilo : travis just gave me timeout running `nipype/pipeline/engine/tests/test_engine.py` https://travis-ci.org/nipy/nipype/jobs/182112718
Not sure if this is the reason, but it was my first push after rebase  current master (after your PR).

**Comment by chrisfilo** [comment]
It's still Python 2.7. Maybe there are some left over MultiProc tests in
test_engine.

**Comment by djarecka** [comment]
@chrisfilo it uses `plugin='Multiproc`: https://github.com/djarecka/nipype/blob/pytest/nipype/pipeline/engine/tests/test_engine.py#L658
Yes, python the same (and the same second run), just want to mention, sine never seen timeout in this test.

**Comment by chrisfilo** [comment]
Ok - we should disable those tests for 2.7 as well (temporarily). Sorry I missed this.

**Comment by djarecka** [comment]
I will do it

**Comment by djarecka** [comment]
@chrisfilo : just that you know there are more tests that uses `plugin="MultiProc`. I can disable them all, but we can also disable ones that cause problems. I don't know what is your suggestions, still don't understand which exactly part leads to the problem.

**Comment by chrisfilo** [comment]
I don't think anyone knows yet :) I would disable all of the tests using
MultiProc on python 2. We should avoid confusing contributors with tests
failing randomly.

**Comment by djarecka** [comment]
@chrisfilo : not sure how to use this information, but it looks like some tests fail more often (the ones that mostly caused problems last week are not the ones that are run first). 
Anyway, I've changed 3 more files: https://github.com/nipy/nipype/pull/1722/commits/fdb71ad7cf4a8a8c7a47625c02eefcb3136760b9

**Comment by chrisfilo** [comment]
The tests fail more often than they did in the past?

**Comment by djarecka** [comment]
no, I can't say this, fortunately don't have enough good statistics... 
I tried to say that for some reason mostly/only `nipype/pipeline/plugins/multiproc.py` was failing last week (at least for me), even though it is not the first test to run with plugin="MultiProc". Don't know if this is important.

**Comment by djarecka** [comment]
the only thing about frequency I can say is that it didn't happen with the master from Oct 19th (I run tests many times, so I'd notice), but I understand that this you know.

**Comment by djarecka** [comment]
@chrisfilo  I was wrong thinking that some tests with `plugin="MultiProc"` don't fail.  Today I run multiple times (in a simple loop)  each test that has  `plugin="MultiProc"` and they all fail from time to time.

**Comment by ccraddock** [comment]
Which tests exactly? We have been doing same loop with test_multiproc and haven't had a problem yet and it's been 6 hours.

**Comment by chrisfilo** [comment]
@djarecka let's disable those tests in python 2 in master ASAP

@djarecka and @ccraddock which exact python versions are you testing with?

**Comment by djarecka** [comment]
@chrisfilo : I did in my pytest PR yesterday, @satra asked me to change Dockerfiles and Circle is running. If I don't have to do anything else, I'd just suggest to wait for the PR to be merged.

**Comment by chrisfilo** [comment]
Ok I see - so this will be merged together with the bit pytest PR. Thank you!

**Comment by djarecka** [comment]
@chrisfilo @ccraddock : running Python 2.7.12, OSX 10.12.1 
running tests `test_run_multiproc`, `test_no_more_memory_than_specifie` from  nipype/pipeline/plugins/tests/test_multiproc.py; `test_serial_input` from nipype/pipeline/engine/tests/test_engine.py; `test_callback_multiproc_normal` and `test_callback_multiproc_exception` from nipype/pipeline/plugins/tests/test_callback.py

I'm running test using pytest from this branch https://github.com/djarecka/nipype/tree/pytest using , I just  removed `@pytest.mark.skipif(sys.version_info < (3, 0)...` for testing purpose.

I usually have problem before running 30th time.

**Comment by djarecka** [comment]
@chrisfilo I should have created a separate PR, but didn't want to rebase again... If my PR will still need some work, I'll create a new PR tomorrow, just to "fix" this issue

**Comment by wtriplett** [comment]
We've seen this deadlock issue as well in production workflows. Some of the cases mentioned previously are with spm, and our workflows also use compiled matlab code (via MCR) unrelated to spm.

To get around this temporarily we've switched to `multiprocessing.semaphore` and give the `acquire()` a 15 sec timeout to keep things moving.

Unfortunately our workflow is really good about reproducing the deadlock so please let me know if I can test anything out!

**Comment by ccraddock** [comment]
I think I have figured this out, https://github.com/nipy/nipype/pull/1754. After making the fixes I ran the nosetests 1000 times without deadlock (or timeouts in the wait()). @wtriplett would you mind seeing if this fixes the problems you are seeing with your workflow?

**Comment by wtriplett** [comment]
@ccraddock: definitely & thanks. Will report back soon.
